unit/getinfo_test.c

```diff
@@ -0,0 +1,226 @@
+		return EXIT_FAILURE;
+	}
+	return 0;
+}
+
+static int check_srcaddr(void *arg)
+{
+	struct fi_info *info = (struct fi_info *)arg;
+	return check_addr(info->src_addr, info->src_addrlen, "source");
+}
+
+static int check_src_dest_addr(void *arg)
+{
+	struct fi_info *info = (struct fi_info *)arg;
+	int ret;
+
+	ret = check_addr(info->src_addr, info->src_addrlen, "source");
+	if (ret)
+		return ret;
+
+	return check_addr(info->dest_addr, info->dest_addrlen, "destination");
+}
+
+int invalid_dom(struct fi_info *hints)
+{
+	if (hints->domain_attr->name)
+		free(hints->domain_attr->name);
+	hints->domain_attr->name = strdup("invalid_domain");
+	if (!hints->domain_attr->name)
+		return -FI_ENOMEM;
+	return 0;
+}
+
+static int getinfo_test(char *node, char *service, uint64_t flags,
+		struct fi_info *hints, ft_getinfo_init init,
+		ft_getinfo_check check, int ret_exp)
+{
+	struct fi_info *info, *fi;
+	int ret;
+
+	if (init) {
+		ret = init(hints);
+		if (ret)
+			return ret;
+	}
+
+	ret = fi_getinfo(FT_FIVERSION, node, service, flags, hints, &info);
+	if (ret) {
+		if (ret == ret_exp)
+			return 0;
+		sprintf(err_buf, "fi_getinfo failed %s(%d)", fi_strerror(-ret), -ret);
+		return ret;
+	}
+
+	if (!check)
+		goto out;
+
+	ft_foreach_info(fi, info) {
+		FT_DEBUG("\nTesting for fabric: %s, domain: %s, endpoint type: %d",
+				fi->fabric_attr->name, fi->domain_attr->name,
+				fi->ep_attr->type);
+		ret = check(info);
+		if (ret)
+			break;
+	}
+out:
+	fi_freeinfo(info);
+	return ret;
+}
+
+#define getinfo_test_addr(number, node, service, flags, hints, init, check,	\
+		ret_exp)							\
+static int getinfo##number(void)						\
+{										\
+	int ret, testret = FAIL;						\
+	ret = getinfo_test(node, service, flags, hints, init, check, ret_exp);	\
+	if (ret)								\
+		goto fail;							\
+	testret = PASS;								\
+fail:										\
+	return TEST_RET_VAL(ret, testret);					\
+}
+
+/*
+ * Tests:
+ */
+
+getinfo_test_addr(1, NULL, NULL, 0, hints, NULL, check_srcaddr, 0)
+getinfo_test_addr(2, NULL, opts.src_port, FI_SOURCE, hints, NULL, check_srcaddr, 0)
+getinfo_test_addr(3, opts.src_addr ? opts.src_addr : "localhost", opts.src_port,
+		FI_SOURCE, hints, NULL, check_srcaddr, 0)
+getinfo_test_addr(4, opts.src_addr ? opts.src_addr : "localhost", NULL,
+		FI_SOURCE, hints, NULL, check_srcaddr, 0)
+getinfo_test_addr(5, opts.dst_addr ? opts.dst_addr : "localhost", opts.dst_port,
+		0, hints, NULL, check_src_dest_addr, 0)
+getinfo_test_addr(6, NULL, NULL, 0, hints, invalid_dom, NULL, -FI_ENODATA)
+
+struct test_entry test_array[] = {
+	TEST_ENTRY(getinfo1, "Test with no node, service or flags"),
+	TEST_ENTRY(getinfo2, "Test with no node, valid service and FI_SOURCE flag"),
+	TEST_ENTRY(getinfo3, "Test with node, valid service and FI_SOURCE flag"),
+	TEST_ENTRY(getinfo4, "Test with node, no service and FI_SOURCE flag"),
+	TEST_ENTRY(getinfo5, "Test with node, service and no flags"),
+	TEST_ENTRY(getinfo6, "Test with non-existent domain name"),
+	{ NULL, "" }
+};
```

Comment:
It would be nice if the description string were more closely associated with the parameters.  For example, move "Test with no node, service or flags" closer to the getinfo_test_addr(1...).  Or maybe redo the getinfo_test_addr macro so it produces a more useful name than getinfo1 -- something like getinfo_no_params, getinfo_src_only, etc.

But overall, I like this addition and would like to add it even if no changes were made.